bacon/model/Tags/TagManager.swift

```diff
@@ -0,0 +1,253 @@
+        let fsm = FileStorageManager()
+        let tagManager: TagManager
+
+        do {
+            log.info("Reading TagManager data from storage.")
+            tagManager = try fsm.readFromJson(TagManager.self, file: TagManager.saveFileName)
+            log.info("TagManager reconstructed from storage data")
+        } catch {
+            // This could happen if a persistent TagManager is instantiated for the first time ever
+            log.warning("\(TagManager.errorLoadData). Instantiating a new instance.")
+            tagManager = TagManager(withPersistence: true)
+            TagManager.cachedPersistentTagManager = tagManager
+            return tagManager
+        }
+
+        log.info("Caching TagManager for future returns.")
+        TagManager.cachedPersistentTagManager = tagManager
+        return tagManager
+    }
+
+    // Private init to disallow direct instantiation: we want to adopt a semi-singleton design pattern.
+    // When TagManager is instantiated in non-persistent mode (e.g. for unit testing),
+    // we can create as many instances as possible.
+    // However, in normal operation (persistent mode), we want to allow only 1 instance,
+    // to avoid multiple working stores of tag data.
+    private init(withPersistence: Bool) {
+        isPersistent = withPersistence
+    }
+
+    func addChildTag(_ child: String, to parent: String) throws {
+        let childTag = Tag(child, parent: parent)
+        let parentTag = Tag(parent)
+
+        // childTag should not already exist
+        guard !allTags.contains(childTag) else {
+            throw DuplicateTagError(message: "\(childTag) already exists")
+        }
+
+        addTags([childTag, parentTag])
+    }
+
+    func addParentTag(_ parent: String) throws {
+        let parentTag = Tag(parent)
+
+        // parentTag should not already exist
+        guard !allTags.contains(parentTag) else {
+            throw DuplicateTagError(message: "\(parentTag) already exists")
+        }
+
+        addTags([parentTag])
+
+    }
+
+    func removeChildTag(_ child: String, from parent: String) throws {
+        let childTag = Tag(child, parent: parent)
+        let parentTag = Tag(parent)
+
+        // childTag should exist
+        guard allTags.contains(childTag) else {
+            throw InvalidTagError(message: "\(childTag) does not exist")
+        }
+
+        // parentTag should exist
+        guard allTags.contains(parentTag) else {
+            throw InvalidTagError(message: "\(parentTag) does not exist")
+        }
+
+        removeTags([childTag])
+    }
+
+    func removeParentTag(_ parent: String) throws {
+        let parentTag = Tag(parent)
+
+        // parentTag should exist
+        guard allTags.contains(parentTag) else {
+            throw InvalidTagError(message: "\(parentTag) does not exist")
+        }
+
+        removeTags([parentTag])
+    }
+
+    var tags: [Tag: Set<Tag>] {
+        return parentChildMap
```

Comment:
I think I might need you to return me a a dictionary where the value is an array of child tags; from the UI side there might need to be a consistent ordering of the child tags under a parent tag

Reply:
This is by design. Within TagManager, I need to do membership check/add/remove operations efficiently, but TagManager does not maintain tags in any order. So an array doesn't make sense.

I think it'll be better if you convert to an array then sort haha

Reply:
As in, you can and should still internally maintain as a set, which will be unknown to me. I'm not sure if I should do any operations (i.e. sorting) on the tags? (separation of concern blah blah)
it's just like transactions internally are not stored in chronological order, but when UI gets hold of it it gets a sorted copy.
but, I'm nor sure haha

Reply:
Okay, done in commit 67f01e5 :)

https://github.com/cs3217-crispybacon/bacon/pull/60/commits/67f01e532c40463a15b253a6ea0d00eb30e2bbfc